### ISSUE TITLE
Shorten method signatures

### ISSUE THREAD

**Issue by bowdenk7** [Microsoft/TypeScript#14662]
Some of the types that make everything work behind the scenes are ridiculous. For example, from Vue.js:

![image](https://cloud.githubusercontent.com/assets/820883/23925649/7be182ca-08cd-11e7-889b-07e907021dd7.png)

For a Vue developer creating a new Vue instance in JavaScript, they will be exposed to this every time. If they already know what arguments to supply, this does not do a great job as a quick reminder. If they don't know what args to supply, this is probably not very helpful either. We should be able to give users an option to show a _concise_ method signature like the following. 

![image](https://cloud.githubusercontent.com/assets/820883/23926202/140d414a-08d0-11e7-9988-1c81f08af533.png)

In this example, the tildas indicate that there is "type madness" hidden below. If the user really wants to see the types, they can goToDefinition or maybe we could give them a way to expand the signature window. 

We would need to pick some arbitrary line when to display type information and when to hide it. My initial thinking is something along the lines of:
- Hide all generic types, but leave angle brackets in
- Hide types when there are more than 2 unions or intersections
- Show string literals if less than 100 characters (roughly 2 lines?)

I would imagine we would also want to give JavaScript users the concise signature by default and TypeScript users the full signature by default, but give both the camps the ability to configure to their preference.

**Comment by mhegazy** [comment]
I would say it should look something like:
![image](https://cloud.githubusercontent.com/assets/8000722/23928010/811592ba-08da-11e7-9097-95b769082b4b.png)

with the definition changed to:

```ts
type VueOptions<D, M, C, P extends string> = ComponentOptions<D, M, C, P[]> & ThisType<D & M & C & Record<P, any> & Vue<D, M, C, P>>;
type VueComponent<D, M, C, P extends string> = D & M & C & Record<P, any> & Vue<D, M, C, Record<P, any>>;
interface Vue<D, M, C, P> { }
interface ComponentOptions<D, M, C, P> { }

declare function vue<D, M, C, P extends string>(options: VueOptions<D, M, C, P>): VueComponent<D, M, C, P>;
```

First shorten the name of the type parameters, and abstract the interesting parts of the type into their own components.

I do not think we should just elide some type parameters just because there is a long type, sometimes these are what you are looking for when you hover over a function/type; i personally do this all the time.

**Comment by RyanCavanaugh** [comment]
Need a formal description of what should happen

**Comment by Kovensky** [comment]
@RyanCavanaugh In the backlog slog, you mention "Encourage people to write type aliases" -- however, type aliases are not used when printing the type in the tooltips. _Only_ `class` and `interface` types are shown by their declared name, but `type` are always replaced by their value.

Compare

`type` | `interface`
--- | ---
 ![image](https://cloud.githubusercontent.com/assets/73085/25884324/7e4f4c04-358c-11e7-9abe-6cadbc66735e.png) | ![image](https://cloud.githubusercontent.com/assets/73085/25884339/96ddaacc-358c-11e7-839c-78bcc440cf88.png)

**Comment by RyanCavanaugh** [comment]
Odd... that should not be happening.
![image](https://cloud.githubusercontent.com/assets/6685088/25912943/e865e720-356d-11e7-9e5f-b7a0591453e4.png)

**Comment by axefrog** [comment]
I agree that unconditional expansion of type aliases is a real problem when writing code. In VS Code, for example:

![image](https://user-images.githubusercontent.com/298883/28494262-3c379ba4-6f6c-11e7-9761-687f4d968f7a.png)

The actual types in question here are:

```ts
export interface SelectorSource<TOut, TIn, TParent extends PotentialValue<TIn>, TSignal>
  extends Source<Selector<TOut>, TIn, TParent, TSignal> {}

export type Selector<TOut> = {
  [P in keyof TOut]: TOut | Selector<TOut[P]> | Source<TOut[P]> | SelectorSource<TOut[P], any, any, any>
};

export type PotentialSource
  <TOut, TIn, TParent extends PotentialValue<TIn>, TSignal>
  = Source<TOut, TIn, TParent, TSignal>
  | SelectorSource<TOut, TIn, TParent, TSignal>;

export type PotentialValue
  <TOut = any, TIn = any, TParent extends PotentialValue<TIn> = any, TSignal = any>
  = TOut
  | Selector<TOut>
  | Source<TOut, TIn, TParent, TSignal>
  | SelectorSource<TOut, TIn, TParent, TSignal>;

export interface HyperscriptHelperFn {
  (): Selector<VNode>;
  (classNameOrId: PotentialValue<string>, data: PotentialValue<VNodeProps>, children: PotentialValue<any[]>): Selector<VNode>;
  (classNameOrId: PotentialValue<string>, data: PotentialValue<VNodeProps>): Selector<VNode>;
  (classNameOrId: PotentialValue<string>, children: PotentialValue<any[]>): Selector<VNode>;
  (classNameOrId: PotentialValue<string>): Selector<VNode>;
  (data: PotentialValue<VNodeProps>): Selector<VNode>;
  (data: PotentialValue<VNodeProps>, children: PotentialValue<any[]>): Selector<VNode>;
  (children: PotentialValue<any[]>): Selector<VNode>;
}
```

TypeScript's type system is a double-edged sword. You can get a lot of power from it, but if you dare to exercise that power, you're punished by unreadable tooltips and stack traces.

My suggestion: for the subsystem within the TypeScript language service that is providing the informational text, provide an option to expand or not expand aliases. In consuming applications, a hotkey or other UI option could then be provided to toggle between expanded and collapsed type alias descriptions as needed, even on a per-function-parameter basis.

**Comment by jpike88** [comment]
Wow this is an old issue. My problem is that I have a function with a param type of `keyof typeof myTranslationKeysJson`, and I get the tokenisation error every time I mouseover it by accident... can we truncate that message?

**Comment by jpike88** [comment]
This is starting to drive me bonkers...

![43304898-04b8bd6e-91a0-11e8-8a94-bf56576cc483](https://user-images.githubusercontent.com/9585787/45134518-1db6e800-b1c5-11e8-84f5-575488cf09d6.png)

@RyanCavanaugh any thoughts on how easy this would be to nail? I wouldn't mind looking into whether I can contribute to this without learning too much, as my time's limited

Keep up the good work guys

**Comment by goldingdamien** [comment]
What is the status of this? The related issues seem to have looked into it, but I can't see any recent developments.

**Comment by tylim88** [comment]
It is almost a year since the last comment, any update?

**Comment by RyanCavanaugh** [comment]
There are no hidden updates.
